File: bench/richards/go/richards.go
// richards — see ../c/richards.c for the algorithm. Task is an interface; each
// task type is a distinct implementer, so run() dispatches dynamically.
package main

import (
	"fmt"
	"os"
	"strconv"
)

const (
	iIdle     = 1
	iWork     = 2
	iHandlerA = 3
	iHandlerB = 4
	iDevA     = 5
	iDevB     = 6
	numIDs    = 6

	kDev  = 1000
	kWork = 1001

	stateRunning           = 0
	stateRunnable          = 1
	stateSuspended         = 2
	stateHeld              = 4
	stateSuspendedRunnable = 3

	dataSize = 4
	count    = 1000
)

type packet struct {
	link *packet
	id   int
	kind int
	a1   int
	a2   [dataSize]int
}

func (p *packet) addTo(list *packet) *packet {
	p.link = nil
	if list == nil {
		return p
	}
	q := list
	for q.link != nil {
		q = q.link
	}
	q.link = p
	return list
}

type tcb struct {
	link     *tcb
	id       int
	priority int
	input    *packet
	state    int
	task     task
}

func (t *tcb) heldOrSuspended() bool {
	return t.state&stateHeld != 0 || t.state == stateSuspended
}

type scheduler struct {
	queueCount int
	holdCount  int
	blocks     [numIDs + 1]*tcb
	list       *tcb
	current    *tcb
	currentID  int
}

func (s *scheduler) hold() *tcb {
	s.holdCount++
	s.current.state |= stateHeld
	return s.current.link
}
func (s *scheduler) suspend() *tcb {
	s.current.state |= stateSuspended
	return s.current
}
func (s *scheduler) release(id int) *tcb {
	t := s.blocks[id]
	if t == nil {
		return nil
	}
	t.state &^= stateHeld
	if t.priority > s.current.priority {
		return t
	}
	return s.current
}
func (s *scheduler) checkPriorityAdd(t *tcb, p *packet) *tcb {
	if t.input == nil {
		t.input = p
		t.state |= stateRunnable
		if t.priority > s.current.priority {
			return t
		}
	} else {
		t.input = p.addTo(t.input)
	}
	return s.current
}
func (s *scheduler) queue(p *packet) *tcb {
	t := s.blocks[p.id]
	if t == nil {
		return nil
	}
	s.queueCount++
	p.link = nil
	p.id = s.currentID
	return s.checkPriorityAdd(t, p)
}
func (s *scheduler) addTask(id, priority int, q *packet, tk task) {
	t := &tcb{link: s.list, id: id, priority: priority, input: q, task: tk}
	if q == nil {
		t.state = stateSuspended
	} else {
		t.state = stateSuspendedRunnable
	}
	s.list = t
	s.current = t
	s.blocks[id] = t
}
func (s *scheduler) schedule() {
	s.current = s.list
	for s.current != nil {
		if s.current.heldOrSuspended() {
			s.current = s.current.link
			continue
		}
		s.currentID = s.current.id
		var p *packet
		if s.current.state == stateSuspendedRunnable {
			p = s.current.input
			s.current.input = p.link
			if s.current.input == nil {
				s.current.state = stateRunning
			} else {
				s.current.state = stateRunnable
			}
		}
		cur := s.current
		s.current = cur.task.run(s, p)
	}
}

type task interface {
	run(s *scheduler, p *packet) *tcb
}

type idleTask struct {
	control, count int
}

func (t *idleTask) run(s *scheduler, p *packet) *tcb {
	t.count--
	if t.count == 0 {
		return s.hold()
	}
	if t.control&1 == 0 {
		t.control >>= 1
		return s.release(iDevA)
	}
	t.control = (t.control >> 1) ^ 53256
	return s.release(iDevB)
}

type deviceTask struct{ pending *packet }

func (t *deviceTask) run(s *scheduler, p *packet) *tcb {
	if p == nil {
		if t.pending == nil {
			return s.suspend()
		}
		q := t.pending
		t.pending = nil
		return s.queue(q)
	}
	t.pending = p
	return s.hold()
}

type workerTask struct{ dest, counter int }

func (t *workerTask) run(s *scheduler, p *packet) *tcb {
	if p == nil {
		return s.suspend()
	}
	if t.dest == iHandlerA {
		t.dest = iHandlerB
	} else {
		t.dest = iHandlerA
	}
	p.id = t.dest
	p.a1 = 0
	for i := 0; i < dataSize; i++ {
		t.counter++
		if t.counter > 26 {
			t.counter = 1
		}
		p.a2[i] = 65 + t.counter - 1
	}
	return s.queue(p)
}

type handlerTask struct{ work, device *packet }

func (t *handlerTask) run(s *scheduler, p *packet) *tcb {
	if p != nil {
		if p.kind == kWork {
			t.work = p.addTo(t.work)
		} else {
			t.device = p.addTo(t.device)
		}
	}
	if t.work != nil {
		c := t.work.a1
		if c < dataSize {
			if t.device != nil {
				v := t.device
				t.device = t.device.link
				v.a1 = t.work.a2[c]
				t.work.a1 = c + 1
				return s.queue(v)
			}
		} else {
			v := t.work
			t.work = t.work.link
			return s.queue(v)
		}
	}
	return s.suspend()
}

func runOnce() (int, int) {
	s := &scheduler{}
	idle := &idleTask{control: 1, count: count}
	s.addTask(iIdle, 0, nil, idle)
	s.blocks[iIdle].state = stateRunning

	wq := (&packet{id: iWork, kind: kWork})
	wq = &packet{link: wq, id: iWork, kind: kWork}
	s.addTask(iWork, 1000, wq, &workerTask{dest: iHandlerA})

	qa := &packet{id: iDevA, kind: kDev}
	qa = &packet{link: qa, id: iDevA, kind: kDev}
	qa = &packet{link: qa, id: iDevA, kind: kDev}
	s.addTask(iHandlerA, 2000, qa, &handlerTask{})

	qb := &packet{id: iDevB, kind: kDev}
	qb = &packet{link: qb, id: iDevB, kind: kDev}
	qb = &packet{link: qb, id: iDevB, kind: kDev}
	s.addTask(iHandlerB, 3000, qb, &handlerTask{})

	s.addTask(iDevA, 4000, nil, &deviceTask{})
	s.addTask(iDevB, 5000, nil, &deviceTask{})

	s.schedule()
	return s.queueCount, s.holdCount
}

func main() {
	iters := 100
	if len(os.Args) > 1 {
		if v, err := strconv.Atoi(os.Args[1]); err == nil {
			iters = v
		}
	}
	q, h := 0, 0
	for i := 0; i < iters; i++ {
		q, h = runOnce()
	}
	fmt.Printf("queue %d hold %d\n", q, h)
}

File: bench/richards/python/richards.py
# richards — see ../c/richards.c for the algorithm.
import sys

I_IDLE, I_WORK, I_HANDLERA, I_HANDLERB, I_DEVA, I_DEVB, NUM_IDS = 1, 2, 3, 4, 5, 6, 6
K_DEV, K_WORK = 1000, 1001
STATE_RUNNING, STATE_RUNNABLE, STATE_SUSPENDED, STATE_HELD, STATE_SUSPENDED_RUNNABLE = 0, 1, 2, 4, 3
DATA_SIZE, COUNT = 4, 1000


class Packet:
    __slots__ = ("link", "id", "kind", "a1", "a2")

    def __init__(self, link, id, kind):
        self.link, self.id, self.kind, self.a1 = link, id, kind, 0
        self.a2 = [0] * DATA_SIZE

    def add_to(self, lst):
        self.link = None
        if lst is None:
            return self
        p = lst
        while p.link is not None:
            p = p.link
        p.link = self
        return lst


class TCB:
    __slots__ = ("link", "id", "priority", "input", "state", "task")

    def __init__(self, link, id, priority, queue, task):
        self.link, self.id, self.priority, self.input, self.task = link, id, priority, queue, task
        self.state = STATE_SUSPENDED if queue is None else STATE_SUSPENDED_RUNNABLE

    def held_or_suspended(self):
        return (self.state & STATE_HELD) != 0 or self.state == STATE_SUSPENDED


class Scheduler:
    def __init__(self):
        self.queue_count = self.hold_count = self.current_id = 0
        self.blocks = [None] * (NUM_IDS + 1)
        self.list = self.current = None

    def hold(self):
        self.hold_count += 1
        self.current.state |= STATE_HELD
        return self.current.link

    def suspend(self):
        self.current.state |= STATE_SUSPENDED
        return self.current

    def release(self, id):
        t = self.blocks[id]
        if t is None:
            return None
        t.state &= ~STATE_HELD
        return t if t.priority > self.current.priority else self.current

    def check_priority_add(self, t, p):
        if t.input is None:
            t.input = p
            t.state |= STATE_RUNNABLE
            if t.priority > self.current.priority:
                return t
        else:
            t.input = p.add_to(t.input)
        return self.current

    def queue(self, p):
        t = self.blocks[p.id]
        if t is None:
            return None
        self.queue_count += 1
        p.link = None
        p.id = self.current_id
        return self.check_priority_add(t, p)

    def add_task(self, id, priority, q, task):
        t = TCB(self.list, id, priority, q, task)
        self.list = self.current = t
        self.blocks[id] = t

    def schedule(self):
        self.current = self.list
        while self.current is not None:
            if self.current.held_or_suspended():
                self.current = self.current.link
                continue
            self.current_id = self.current.id
            p = None
            if self.current.state == STATE_SUSPENDED_RUNNABLE:
                p = self.current.input
                self.current.input = p.link
                self.current.state = STATE_RUNNING if self.current.input is None else STATE_RUNNABLE
            cur = self.current
            self.current = cur.task.run(self, p)


class IdleTask:
    __slots__ = ("control", "count")

    def __init__(self):
        self.control, self.count = 1, COUNT

    def run(self, s, p):
        self.count -= 1
        if self.count == 0:
            return s.hold()
        if self.control & 1 == 0:
            self.control >>= 1
            return s.release(I_DEVA)
        self.control = (self.control >> 1) ^ 53256
        return s.release(I_DEVB)


class DeviceTask:
    __slots__ = ("pending",)

    def __init__(self):
        self.pending = None

    def run(self, s, p):
        if p is None:
            if self.pending is None:
                return s.suspend()
            q = self.pending
            self.pending = None
            return s.queue(q)
        self.pending = p
        return s.hold()


class WorkerTask:
    __slots__ = ("dest", "counter")

    def __init__(self):
        self.dest, self.counter = I_HANDLERA, 0

    def run(self, s, p):
        if p is None:
            return s.suspend()
        self.dest = I_HANDLERB if self.dest == I_HANDLERA else I_HANDLERA
        p.id = self.dest
        p.a1 = 0
        for i in range(DATA_SIZE):
            self.counter += 1
            if self.counter > 26:
                self.counter = 1
            p.a2[i] = 65 + self.counter - 1
        return s.queue(p)


class HandlerTask:
    __slots__ = ("work", "device")

    def __init__(self):
        self.work = self.device = None

    def run(self, s, p):
        if p is not None:
            if p.kind == K_WORK:
                self.work = p.add_to(self.work)
            else:
                self.device = p.add_to(self.device)
        if self.work is not None:
            c = self.work.a1
            if c < DATA_SIZE:
                if self.device is not None:
                    v = self.device
                    self.device = self.device.link
                    v.a1 = self.work.a2[c]
                    self.work.a1 = c + 1
                    return s.queue(v)
            else:
                v = self.work
                self.work = self.work.link
                return s.queue(v)
        return s.suspend()


def run_once():
    s = Scheduler()
    s.add_task(I_IDLE, 0, None, IdleTask())
    s.blocks[I_IDLE].state = STATE_RUNNING
    wq = Packet(None, I_WORK, K_WORK)
    wq = Packet(wq, I_WORK, K_WORK)
    s.add_task(I_WORK, 1000, wq, WorkerTask())
    qa = Packet(None, I_DEVA, K_DEV)
    qa = Packet(qa, I_DEVA, K_DEV)
    qa = Packet(qa, I_DEVA, K_DEV)
    s.add_task(I_HANDLERA, 2000, qa, HandlerTask())
    qb = Packet(None, I_DEVB, K_DEV)
    qb = Packet(qb, I_DEVB, K_DEV)
    qb = Packet(qb, I_DEVB, K_DEV)
    s.add_task(I_HANDLERB, 3000, qb, HandlerTask())
    s.add_task(I_DEVA, 4000, None, DeviceTask())
    s.add_task(I_DEVB, 5000, None, DeviceTask())
    s.schedule()
    return s.queue_count, s.hold_count


def main():
    iters = int(sys.argv[1]) if len(sys.argv) > 1 else 100
    q = h = 0
    for _ in range(iters):
        q, h = run_once()
    print("queue %d hold %d" % (q, h))


main()
